04-frequent-scenarios.md

```diff
@@ -0,0 +1,288 @@
+# Frequent scenarios
+
+## Introduction
+
+Many use cases can be captured in a small set of scenarios. This chapter aims to illustrate this small set and get you up and running as soon as possible with the REST convention. 
+
+## Scenarios
+
+The scenarios described center around the [HTTP Response Status Codes][http-status-codes]. 
+
+We will use the following codes:
+* [200](#200--happy-flow); the expected response, ie the happy flow scenario
+* [301](#301--redirected); request for the client to retrieve the resource from another location
+* [304](#304--apply-cache); request for the client to utilize its cache
+* [400](#400--validation-error); an error was encountered that can be fixed by the client
+* [401](#401--unauthenticated); the client has not been authenticated
+* [403](#403--unauthorized); the client tried to access a resource for which it was not authorized
+* [404](#404--not-found); the requested resource could not be found
+* [500](#500--critical-error); the server encountered an error from which it cannot recover
+
+### 200 / Happy flow
+
+Typical response status code: 200
+
+Request
+```
+GET /cars/1
+```
+
+Response
+```
+Status code: 200
+Content: <JSON body of cars instance>
+```
+
+```java
+@RestController
+@RequestMapping("/cars")
+public class CarController {
+
+    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
+    public Car getCar(@PathVariable Long id) {
+        return carService.findOne(id);
+    }
+}
+```
+
+If the car resource is found, Spring will automatically make sure a response is created with a 200 code. You will have to do nothing special for this to work.
+
+### 301 / Redirected
+
+A redirection takes place if the server can determine whether a resource has been moved to another location, and also knows what that new location is. A typical example of valid usage here, is a resource which is addressed by a human-readable permalink (for example, the title of an article). When the title changes, the resource has effectively moved. This situation warrants a redirect operation.
+
+Ideally, if a resource has been moved to another link, the client will be told so in the standard HTTP way. In this case, the server method responsible for looking up the resource will throw an exception that is handled by a @ControllerAdvice method. This method will set the response code (301) and add the header with the new location of the resource.
+
+```java
+@ExceptionHandler(CarRedirectException.class)
```

Comment:
Ah right, I misread the sentence thinking that the endpoint of all cars had moved (which I think would be controller logic than business logic) but this is about a specific car for which the permalink URL has changed. In that case you would indeed require business logic to know that the URL has changed so throwing an Exception in the service makes sense. :)

Regarding handling the exceptions in a ControllerAdvice, I completely agree. The only thing I was wondering about if in this specific situation we'd require an exception because I misunderstood the example. 

+1